Support balances and add warnings

treydock opened the pull request:
Screenshot using default locales

<img width="1331" alt="Screen Shot 2019-08-30 at 3 15 49 PM" src="https://user-images.githubusercontent.com/739622/64046166-2cff1600-cb39-11e9-807e-4f412d3d58ae.png">

Screenshot using OSC locales

<img width="1349" alt="Screen Shot 2019-08-30 at 3 13 36 PM" src="https://user-images.githubusercontent.com/739622/64046156-27a1cb80-cb39-11e9-96de-44ca9e933c94.png">

ericfranz commented on app/models/balance.rb:
```diff
@@ -0,0 +1,87 @@
+# This describes disk quota utilization for a given user and volume
```
Update comment to describe balance instead of disk quota. Maybe in parentheses (RUs) in comments for us on Gateways?

MorganRodgers replied:
I would prefer that the comment refer to `config.unit` specifically, and call out resource units as what are used at OSC.

treydock replied:
Shouldn't that go in README or does it belong in model header too?

ericfranz commented on app/models/balance.rb:
```diff
@@ -0,0 +1,87 @@
+
+  def balance_object
+    return @project if @project.present?
+    @user
```
Can use the new [presence](https://guides.rubyonrails.org/active_support_core_extensions.html#presence) here:

```ruby
def balance_object
  project.presence || user
end
```

treydock replied:
Change made.

ericfranz commented on config/configuration_singleton.rb:
```diff
@@ -109,6 +109,25 @@ def quota_threshold
+  # The threshold for determining if there is sufficient balance remaining
+  # @return [Float] threshold factor
+  def balance_threshold
+    ENV.fetch("OOD_BALANCE_THRESHOLD", 0).to_f
```
The default threshold should be higher than 0. If 0, the warning message will always display, if this works like the quota warnings do.

ericfranz replied:
Sorry I realize now that the threshold value is an integer i.e. number of RUs, rather than a percentage. Markign this as resolved.

ericfranz commented on README.md:
````diff
@@ -189,6 +189,77 @@ be repeated with just `user`, `block_usage`, and `file_usage` changing.*
+```
+
+Where `version` defines the version of the JSON schema used, `timestamp`
+defines when this file was generated, and `quotas` is a list of quota objects
````
you mean `balances` not `quotas`

treydock replied:
Fixed copy+paste mistake

ericfranz commented on README.md:
```diff
@@ -189,6 +189,77 @@ be repeated with just `user`, `block_usage`, and `file_usage` changing.*
+environment variable `OOD_BALANCE_PATH` as a colon-delimited list of all JSON
+file paths.
+
+The default threshold for displaying the warning is at `0`, but this
```
Should the default be something other than 0?

ericfranz replied:
Oh I see threshold is not a percentage but an integer value like 1000.

MorganRodgers commented on README.md:
```diff
@@ -189,6 +189,77 @@ be repeated with just `user`, `block_usage`, and `file_usage` changing.*
+defines when this file was generated, and `quotas` is a list of quota objects
+(see below).
+
+The value for `config.unit` defines the type of units for balances and `config.project_type` would be project, account, or group, etc.
```
You should explicitly state that this value is used in the locale files, and so can be any string value that makes sense.

treydock replied:
Updated to call both out.

treydock commented:
I updated it so `config` key is optional and added better default locale text if no units are provided

<img width="569" alt="Screen Shot 2019-09-03 at 1 54 26 PM" src="https://user-images.githubusercontent.com/739622/64196800-a147fb00-ce52-11e9-9f6a-77c5356572ec.png">

treydock commented:
Slightly better:

<img width="456" alt="Screen Shot 2019-09-03 at 2 01 25 PM" src="https://user-images.githubusercontent.com/739622/64197160-567ab300-ce53-11e9-8192-297f2548ae0d.png">

treydock commented:
Turns out updating every 15 minutes is useless because OSC only sends charging data to database once a day, which seems logical other sites do things similarly as easier to charge for previous day than several times a day.  Going to remove the refresh locale.

treydock commented:
Actually might leave text but modify for default to daily.

treydock commented:
One thing I added and can revert if not wanted is concept of OnDemand having timezone configured. The ActiveSupport time zone names I couldn't get from local server timezone via `Time` so had to configure the Rails app. I couldn't set time zone with initializer since has to be set in `Application` class. The absence of a timezone just sticks with UTC which is the default. Since the time used for things like balances is specific to where the server is deployed and not the user's timzone, I opted for server-side time zone configuration.

treydock commented:
This is current look using OSC specific units and project type but default locales.

<img width="1233" alt="Screen Shot 2019-09-04 at 8 04 14 AM" src="https://user-images.githubusercontent.com/739622/64253148-9e98e480-ceea-11e9-86b1-8da9b7143848.png">

ericfranz commented on config/configuration_singleton.rb:
```diff
@@ -109,6 +109,25 @@ def quota_threshold
+  # @return [Array<String>] balance paths
+  def balance_paths
+    # regex uses negative lookahead to ignore : preceeding //
+    ENV.fetch("OOD_BALANCE_PATH", "").strip.split(/:(?!\/\/)/)
```
we should probably memoize this and `quota_paths` since this is executed on every request...

```ruby
def balance_paths
  @balance_paths ||= ENV.fetch("OOD_BALANCE_PATH", "").strip.split(/:(?!\/\/)/)
end
```

treydock replied:
I made the change but is looking up something in the environment and doing a few operations on the value really that expensive?

ericfranz replied:
No its not. But its also easy to memoize. I guess we didn't do that before because we didn't actively memoize everything we could. Maybe its a bad idea in practice to do that there and instead utilize view caching where appropriate. For example, parsing quota files is far more expensive than getting the list of files to parse by using a regex.

ericfranz replied:
You know what adding that memoization was probably bad advice. I vaguely remember avoiding or removing excessive memoization in this file for this reason. If you want to revert you can.

ericfranz commented:
We should actually revert the timezone change. In the Job Composer we use this Rails gem https://github.com/basecamp/local_time. That adds a view helper "local_time" so you can do something [like this](https://github.com/OSC/ood-myjobs/blob/41cb76814f28387085e419eebe78ea472ed1d99d/app/views/workflows/index.html.erb#L84):

```
<%= local_time(workflow.created_at) %>
```

The result is the user sees something in their own time zone instead of UTC time formatted string. Thus, if I create a new job at Utah's CHPC or at OSC in the Job Composer the created at displays like this: "September 6, 2019 2:35pm"  - even though the time zones are different I know as a user I created both of those 15 minutes ago since it is 2:50 PM right now.

If you want to display this string in local time that would be the most appropriate to do now.

As far as having an OOD_TIMEZONE, you could open a separate issue on the ondemand repo about the topic of displaying time in OnDemand and we could adopt a more formalized strategy that addresses drawbacks or insufficiency of using Basecamp's local_time gem.

treydock commented:
Added commit that uses `local_time` to generate the last_update value.

ericfranz commented on app/views/shared/_insufficient_balance_resource.html.erb:
```diff
@@ -2,7 +2,7 @@
   <strong><%= balance.project_type %> <%= balance.balance_object %></strong>
   <small class="pull-right hidden-xs hidden-sm">
-    <%= t('dashboard.balance_reload_message_html', last_update: balance.last_update) %>
+    <%= t('dashboard.balance_reload_message_html', last_update: local_time(balance.last_update)) %>
```
This could just be:

```
    <%= t('dashboard.balance_reload_message_html', last_update: local_time(balance.updated_at)) %>
```

and `Balance#last_update` removed as `Balance#last_update` purpose is to make a local time human readable form of `Balance#updated_at`, which is what `local_time(balance.updated_at)` also achieves.

ericfranz commented on app/views/shared/_insufficient_balance_resource.html.erb:
```diff
@@ -0,0 +1,12 @@
+<h4>
+  <%= t('dashboard.balance_warning_prefix_html', units_balance: balance.units_balance, unit: @unit) %>
```
`@unit` is an instance variable of `Balance` so should this instead be:

```
  <%= t('dashboard.balance_warning_prefix_html', units_balance: balance.units_balance, unit: balance.unit) %>
```

ericfranz commented on app/views/shared/_insufficient_balance_resource.html.erb:
```diff
@@ -0,0 +1,12 @@
+</h4>
+<ul class="list-unstyled">
+  <li>
+    <%= balance.to_s %>. <%= t('dashboard.balance_additional_message', balanace_units: balance.balanace_units, unit: @unit) %>
```
See comment above, it seems like `@unit` here should actually be `balance.unit`
